storage/innobase/include/arch0arch.h

```diff
@@ -0,0 +1,1984 @@
+};
+
+/** Archived data block state.
+A data block is a block in memory that holds dirty page IDs before persisting
+into disk. Shown below is the state transfer diagram for a data block.
+
+@startuml
+
+  state ARCH_BLOCK_INIT
+  state ARCH_BLOCK_ACTIVE
+  state ARCH_BLOCK_READY_TO_FLUSH
+  state ARCH_BLOCK_FLUSHED
+
+  [*] -down-> ARCH_BLOCK_INIT
+  ARCH_BLOCK_INIT -> ARCH_BLOCK_ACTIVE : Writing page ID
+  ARCH_BLOCK_ACTIVE -> ARCH_BLOCK_READY_TO_FLUSH : Block is full
+  ARCH_BLOCK_READY_TO_FLUSH -> ARCH_BLOCK_FLUSHED : Block is flushed
+  ARCH_BLOCK_FLUSHED --> ARCH_BLOCK_ACTIVE : Writing page ID
+  ARCH_BLOCK_FLUSHED -down-> [*]
+
+@enduml */
+enum Arch_Blk_State
+{
+  /** Data block is initialized */
+  ARCH_BLOCK_INIT = 0,
+
+  /** Data block is active and having data */
+  ARCH_BLOCK_ACTIVE,
+
+  /** Data block is full but not flushed to disk */
+  ARCH_BLOCK_READY_TO_FLUSH,
+
+  /** Data block is flushed and can be reused */
+  ARCH_BLOCK_FLUSHED
+};
+
+/** Archiver block type */
+enum Arch_Blk_Type
+{
+  /* Block which holds reset information */
+  ARCH_RESET_BLOCK = 0,
+
+  /* Block which holds archived page IDs */
+  ARCH_DATA_BLOCK
+};
+
+/** Archiver block flush type */
+enum Arch_Blk_Flush_Type
+{
+  /** Flush when block is full */
+  ARCH_FLUSH_NORMAL = 0,
+
+  /** Flush partial block.
+  Needed for persistent page tracking. */
+  ARCH_FLUSH_PARTIAL
+};
+
+/** Page Archive doublewrite buffer block offsets */
+enum Arch_Page_Dblwr_Offset
+{
+  /** Archive doublewrite buffer page offset for RESET page. */
+  ARCH_PAGE_DBLWR_RESET_PAGE = 0,
+
+  /* Archive doublewrite buffer page offset for FULL FLUSH page. */
+  ARCH_PAGE_DBLWR_FULL_FLUSH_PAGE,
+
+  /* Archive doublewrite buffer page offset for PARTIAL FLUSH page. */
+  ARCH_PAGE_DBLWR_PARTIAL_FLUSH_PAGE
+};
+
+/** Forward declarations */
+class Arch_Group;
+class Arch_Log_Sys;
+class Arch_Dblwr_Ctx;
+class Arch_Recv_Group_Info;
+
+/** Guard to release resources safely */
+class Arch_scope_guard
+{
+ public:
+  /** Attach a function to the guard which releases some resource. */
+  Arch_scope_guard(std::function<void()> function) { m_cleanup = function; }
+
+  /** Release the resources automatically at the time of destruction. */
+  ~Arch_scope_guard()
+  {
+    if (m_cleanup)
+      m_cleanup();
+  }
+
+  /** Manually release the resource. */
+  void cleanup()
+  {
+    m_cleanup();
+    m_cleanup = nullptr;
+  }
+
+ private:
+  /** Function to release the resource. */
+  std::function<void()> m_cleanup{};
+};
+
+/** Position in page ID archiving system */
+struct Arch_Page_Pos
+{
+  /** Initialize a position */
+  void init();
+
+  /** Position in the beginning of next block */
+  void set_next();
+
+  /** Unique block number */
+  uint32_t m_block_num;
+
+  /** Offset within a block */
+  uint m_offset;
+
+  bool operator<(Arch_Page_Pos pos)
+  {
+    return (m_block_num < pos.m_block_num ||
+            (m_block_num == pos.m_block_num && m_offset <= pos.m_offset));
+  }
+};
```

Comment:
This had better be a simple wrapper on top of a single `uint64_t`, similar to `page_id_t`. Then, any comparisons as well as `-=` and `+=` would be free of any conditions in non-debug builds. We would only have debug assertions to prevent some overflow, like this:
```c++
  page_id_t operator-(uint32_t i) const noexcept
  {
    ut_ad(page_no() >= i);
    return page_id_t(m_id - i);
  }
  page_id_t operator+(uint32_t i) const noexcept
  {
    ut_ad(page_no() < ~i);
    return page_id_t(m_id + i);
  }
```